src/docs/arch-logging.dox

```diff
@@ -1,8 +1,167 @@
+logging subsystem allocates a 128-byte multiple rounded up buffer and copies the
+record into that buffer and zero fills the tail.
+
+Logging supports both compression and at-rest encryption. When compression is enabled,
+WiredTiger only compresses if doing so may result in space savings. Therefore, small records
+that already are below the 128 byte minimum size will not result in any space savings.
+All user log records are encrypted when enabled, regardless of size. When both
+compression and encryption are enabled, compression is performed first, then encryption.
+That resulting log record is then rounded up to the 128-byte multiple.
+
+WiredTiger uses a lock free mechanism to write records to the log
+in the common path.  WiredTiger has a pool of log buffers called
+slots and each thread entering the logging code reserves its memory
+buffer via atomic operations and then fills in its part of the
+memory. When all threads that are using the slot finish their copies
+into the memory, then the slot can be released and written to the
+file system. Slots remain open and the memory available for the
+next thread to use until a flush is forced, the memory fills or a
+timer expires.
+
+This means that for the common case of a user committing a transaction with the
+default synchronization setting of memory-only the path of writing its record into
+the write-ahead log becomes essentially two atomic operations and memory copies. The
+common path does not require acquiring locks.
+
+@subsection Internal Threads
+There are several internal logging-related threads that perform housekeeping on the
+logging subsystem. The main thread, called \c log_server, is a general thread. It
+flushes out the current slot to the file system. It does this every 50 milliseconds
+and this is done so that if a system becomes idle, buffered log records in the memory
+of a slot do not languish. The thread also performs log archiving to remove any log
+files that are no longer needed based on the checkpoint LSN. Finally the main thread
+performs log file pre-allocation, if configured. The pre-allocation algorithm is
+adaptable. If the thread did not make sufficient numbers of pre-allocated log files
+the system then makes more. If too few were used, then later pre-allocation will
+create fewer of them.
+
+There is another internal thread dedicated to managing operations on log files. This
+is the \c log_file_server thread. I/O operations can be slow and costly so this
+thread takes the burden of closing and syncing files to disk off of user application threads.
+If a log file needs to be closed, then the application thread moves its file handle to the
+\c log->log_close_fh field and signals this thread. This thread will fsync the file
+handle, and then close it. Updating the \c log_close_lsn and the \c sync_lsn values.
+
+The third internal thread manages advancing the \c write_lsn in proper LSN order.
```

Comment:
What does proper LSN order mean? I feel like the second sentence repeats what you said in the first. Would it make sense to rewrite as "The third internal thread manages advancing the `write_lsn`. As the `write_lsn` must be advanced in proper LSN order, it becomes a point of single threading in the path of writing a log record."